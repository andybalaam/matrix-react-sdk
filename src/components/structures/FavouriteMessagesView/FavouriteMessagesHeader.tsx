/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { FC, useCallback, useState } from "react";

import { Action } from "../../../dispatcher/actions";
import defaultDispatcher from "../../../dispatcher/dispatcher";
import useFavouriteMessages from "../../../hooks/useFavouriteMessages";
import { _t } from "../../../languageHandler";
import { FavouriteMessagesStore } from "../../../stores/FavouriteMessagesStore";
import RoomAvatar from "../../views/avatars/RoomAvatar";
import AccessibleTooltipButton from "../../views/elements/AccessibleTooltipButton";

interface IProps {
    query: string;
    handleSearchQuery: (query: string) => void;

    // Provide this to use a different store for favourite messages
    // e.g. in tests. If not supplied, we use the global default.
    favouriteMessagesStore?: FavouriteMessagesStore;
}

const FavouriteMessagesHeader: FC<IProps> = ({ query, handleSearchQuery, favouriteMessagesStore }: IProps) => {
    const { getFavouriteMessages } = useFavouriteMessages(favouriteMessagesStore);
    const favouriteMessagesIds = getFavouriteMessages();

    const [isSearchClicked, setSearchClicked] = useState<boolean>(false);

    const onChange = useCallback((e) => handleSearchQuery(e.target.value), [handleSearchQuery]);
    const onCancelClick = useCallback(() => {
        setSearchClicked(false);
        handleSearchQuery("");
    }, [handleSearchQuery, setSearchClicked]);
    const onSearchClick = useCallback(() => setSearchClicked(true), [setSearchClicked]);

    const onClearClick = useCallback(() => {
        if (favouriteMessagesIds.length > 0) {
            defaultDispatcher.dispatch({ action: Action.OpenClearFavourites });
        }
    }, [favouriteMessagesIds]);

    return (
        <div className="mx_FavMessagesHeader">
            <div className="mx_FavMessagesHeader_Wrapper">
                <div className="mx_FavMessagesHeader_Wrapper_left">
                    <RoomAvatar
                        oobData={{
                            name: "Favourites",
                        }}
                        width={26}
                        height={26}
                        resizeMethod="crop"
                    />
                    <span>Favourite Messages</span>
                </div>
                <div className="mx_FavMessagesHeader_Wrapper_right">
                    {isSearchClicked ? (
                        <>
                            <input
                                type="text"
                                className="mx_FavMessagesHeader_Search"
                                placeholder="Search..."
                                onChange={onChange}
                                value={query}
                            />
                            <AccessibleTooltipButton
                                className="mx_FavMessagesHeader_cancelButton"
                                onClick={onCancelClick}
                                title={_t("Cancel")}
                                key="cancel"
                            />
                        </>
                    ) : (
                        <AccessibleTooltipButton
                            className="mx_RoomHeader_button mx_RoomHeader_searchButton"
                            onClick={onSearchClick}
                            title={_t("Search")}
                            key="search"
                        />
                    )}
                    <AccessibleTooltipButton
                        className="mx_RoomHeader_button mx_FavMessagesHeader_clearAllButton"
                        onClick={onClearClick}
                        title={_t("Clear")}
                        key="clear"
                    />
                </div>
            </div>
        </div>
    );
};

export default FavouriteMessagesHeader;
