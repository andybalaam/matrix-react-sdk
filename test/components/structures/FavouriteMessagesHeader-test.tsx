/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render } from "@testing-library/react";

import { stubClient } from "../../test-utils";
import SettingsStore from "../../../src/settings/SettingsStore";
import FavouriteMessagesHeader from "../../../src/components/structures/FavouriteMessagesView/FavouriteMessagesHeader";

describe("FavouriteMessagesHeader", () => {
    beforeEach(async () => {
        jest.spyOn(SettingsStore, "getValue").mockImplementation((setting) => setting === "feature_favourite_messages");
        stubClient();
    });

    afterEach(async () => {
        jest.resetAllMocks();
    });

    it("displays a title and buttons", async () => {
        const view = render(<FavouriteMessagesHeader query="" handleSearchQuery={() => {}} />);
        view.getByTestId("avatar-img");
        view.getByText("Favourite Messages");
        view.getByLabelText("Search");
        view.getByLabelText("Clear");
        expect(view.asFragment()).toMatchSnapshot();
    });

    it("displays a search box after Search is clicked", async () => {
        // Given a favourites header with a (hidden) search query
        const view = render(<FavouriteMessagesHeader query="foo" handleSearchQuery={() => {}} />);
        expect(view.queryByRole("textbox")).toBeNull();

        // When we click Search
        view.getByLabelText("Search").click();

        // Then the search box appears
        const textbox = view.getByRole("textbox");
        // And it contains our search query
        expect(textbox.getAttribute("placeholder")).toBe("Search...");
        expect(textbox.getAttribute("value")).toBe("foo");

        expect(view.asFragment()).toMatchSnapshot();
    });

    it("hides the search box when you click Cancel", async () => {
        // Given a favourites header where Search has been clicked
        const view = render(<FavouriteMessagesHeader query="" handleSearchQuery={() => {}} />);
        expect(view.queryByRole("textbox")).toBeNull();
        view.getByLabelText("Search").click();

        // Sanity: Search button has disappeared and textbox has appeared
        expect(view.queryByLabelText("Search")).toBeNull();
        view.getByRole("textbox");

        // When we click Cancel
        view.getByLabelText("Cancel").click();

        // Then the search box disappeared
        expect(view.queryByRole("textbox")).toBeNull();
        // And the Cancel button transformed back into Search
        expect(view.queryByLabelText("Cancel")).toBeNull();
        view.getByLabelText("Search");
    });
});
