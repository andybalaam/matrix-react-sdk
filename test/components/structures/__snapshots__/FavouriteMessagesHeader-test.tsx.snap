// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouriteMessagesHeader displays a search box after Search is clicked 1`] = `
<DocumentFragment>
  <div
    class="mx_FavMessagesHeader"
  >
    <div
      class="mx_FavMessagesHeader_Wrapper"
    >
      <div
        class="mx_FavMessagesHeader_Wrapper_left"
      >
        <span
          class="mx_BaseAvatar"
          role="presentation"
        >
          <span
            aria-hidden="true"
            class="mx_BaseAvatar_initial"
            style="font-size: 16.900000000000002px; width: 26px; line-height: 26px;"
          >
            F
          </span>
          <img
            alt=""
            aria-hidden="true"
            class="mx_BaseAvatar_image"
            data-testid="avatar-img"
            src="data:image/png;base64,00"
            style="width: 26px; height: 26px;"
          />
        </span>
        <span>
          Favourite Messages
        </span>
      </div>
      <div
        class="mx_FavMessagesHeader_Wrapper_right"
      >
        <input
          class="mx_FavMessagesHeader_Search"
          placeholder="Search..."
          type="text"
          value="foo"
        />
        <div
          aria-label="Cancel"
          class="mx_AccessibleButton mx_FavMessagesHeader_cancelButton"
          role="button"
          tabindex="0"
        />
        <div
          aria-label="Clear"
          class="mx_AccessibleButton mx_RoomHeader_button mx_FavMessagesHeader_clearAllButton"
          role="button"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FavouriteMessagesHeader displays a title and buttons 1`] = `
<DocumentFragment>
  <div
    class="mx_FavMessagesHeader"
  >
    <div
      class="mx_FavMessagesHeader_Wrapper"
    >
      <div
        class="mx_FavMessagesHeader_Wrapper_left"
      >
        <span
          class="mx_BaseAvatar"
          role="presentation"
        >
          <span
            aria-hidden="true"
            class="mx_BaseAvatar_initial"
            style="font-size: 16.900000000000002px; width: 26px; line-height: 26px;"
          >
            F
          </span>
          <img
            alt=""
            aria-hidden="true"
            class="mx_BaseAvatar_image"
            data-testid="avatar-img"
            src="data:image/png;base64,00"
            style="width: 26px; height: 26px;"
          />
        </span>
        <span>
          Favourite Messages
        </span>
      </div>
      <div
        class="mx_FavMessagesHeader_Wrapper_right"
      >
        <div
          aria-label="Search"
          class="mx_AccessibleButton mx_RoomHeader_button mx_RoomHeader_searchButton"
          role="button"
          tabindex="0"
        />
        <div
          aria-label="Clear"
          class="mx_AccessibleButton mx_RoomHeader_button mx_FavMessagesHeader_clearAllButton"
          role="button"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
